File: app/src/main/java/com/example/fooddeuk/util/RealmUtil.java
package com.example.fooddeuk.util;

import com.example.fooddeuk.model.cart.CartOption;
import com.orhanobut.logger.Logger;

import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.RealmResults;

/**
 * Created by heojuyeong on 2017. 9. 19..
 */

public class RealmUtil {
    static Realm realm=Realm.getDefaultInstance();

    public static <T extends RealmObject> T getRealmObject(final T data){

        try{
            realm.beginTransaction();
            T realmResult=realm.copyToRealm(data);
            realm.commitTransaction();
            return realmResult;
        }catch (Throwable e){
            if(realm.isInTransaction()) {
                realm.cancelTransaction();
            }
            e.printStackTrace();
        }
        return null;

    }
    public static <T extends RealmObject> int getDataSize(Class<T> clazz){

        try{
            realm.beginTransaction();
            int size= realm.where(clazz).findAll().size();
            realm.commitTransaction();
            return size;

        }catch (Throwable e){
            if(realm.isInTransaction()) {
                realm.cancelTransaction();
            }
            e.printStackTrace();
        }
        return 0;
    }

    public static <T extends RealmObject>void  insertData(final T data){

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.copyToRealmOrUpdate(data);
            }
        });
    }



    public static <T extends RealmObject>void  insertData2(final T data){

        realm.beginTransaction();
        realm.insertOrUpdate(data);
        realm.commitTransaction();
    }

    public static void test(){

        RealmResults<CartOption> cartOption=realm.where(CartOption.class).findAll();
        CartOption ca=cartOption.first();
        int op=cartOption.size();
        Logger.d(op);
        Logger.d(ca);

    }

    public static <T extends RealmObject> RealmResults<T> findDataAll(Class<T> clazz) {
        try{
            realm.beginTransaction();
            RealmResults<T> item = realm.where(clazz).findAll();
            realm.commitTransaction();
            return item;

        }catch (Throwable e){
            if(realm.isInTransaction()) {
                realm.cancelTransaction();
            }
            e.printStackTrace();
        }

        return null;
    }

    public static <T extends RealmObject> T findData(Class<T> clazz) {
        try{
            realm.beginTransaction();
            T item = realm.where(clazz).findFirst();
            realm.commitTransaction();
            return item;

        }catch (Throwable e){
            if(realm.isInTransaction()) {
                realm.cancelTransaction();
            }
            e.printStackTrace();
        }

        return null;
    }

    public static <T extends RealmObject> RealmResults<T> findDataById(Class<T> clazz, int id){
        RealmResults<T> item=realm.where(clazz).equalTo("id",id).findAll();
        return item;
    }

    public static <T extends RealmObject>void removeDataAll(Class<T> clazz){
        final RealmResults<T> result = findDataAll(clazz);
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                result.deleteAllFromRealm();
            }
        });
    }

    public static <T extends RealmObject>void removeDataById(Class<T> clazz, int id){
        final RealmResults<T> result = findDataById(clazz,id);
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                result.deleteAllFromRealm();
            }
        });
    }



    //Auto Increment id
    public static<T extends RealmObject> int getAutoIncrementId(Class<T> clazz) {
        try {
            Number id = realm.where(clazz).max("id");
            if (id != null) {
                return id.intValue() + 1;
            } else {
                return 0;
            }
        } catch (ArrayIndexOutOfBoundsException e) {
            return 0;
        }
    }






}

File: app/src/main/java/com/example/fooddeuk/util/RealmUtil.kt
package com.example.fooddeuk.util

import com.example.fooddeuk.`object`.GlobalApplication
import io.realm.Realm
import io.realm.RealmObject
import io.realm.RealmResults

/**
 * Created by heojuyeong on 2017. 9. 19..
 */

object RealmUtil {
    val realm: Realm = GlobalApplication.getInstance().realm
    val inRealm: Realm = GlobalApplication.getInstance().inRealm

    fun beginTranscation(callback : () -> Boolean){

        try {
            realm.beginTransaction()
            if(callback()){
                realm.commitTransaction()
            }

        } catch (e: Throwable) {
            if (realm.isInTransaction) {
                realm.cancelTransaction()
            }
            e.printStackTrace()
        }
    }
    

    fun <T : RealmObject> getRealmObject(data: T): T? {

        try {

            realm.beginTransaction()
            val realmResult = realm.copyToRealm(data)
            realm.commitTransaction()
            return realmResult
        } catch (e: Throwable) {
            if (realm.isInTransaction) {
                realm.cancelTransaction()
            }
            e.printStackTrace()
        }

        return null

    }

    fun <T : RealmObject> getDataSize(clazz: Class<T>): Int {

        try {

            realm.beginTransaction()
            val size = realm.where(clazz).findAll().size
            realm.commitTransaction()
            return size

        } catch (e: Throwable) {
            if (realm.isInTransaction) {
                realm.cancelTransaction()
            }
            e.printStackTrace()
        }

        return 0
    }

    fun <T : RealmObject> insertData(data: T) {

        realm.executeTransaction { realm -> realm.copyToRealmOrUpdate(data) }
    }


    fun <T : RealmObject> insertInData(data: T) {

        inRealm.executeTransaction { realm -> realm.copyToRealmOrUpdate(data) }
    }


    fun <T : RealmObject> findDataAll(clazz: Class<T>): RealmResults<T>? {

        try {
            realm.beginTransaction()
            val item = realm.where(clazz).findAll()
            realm.commitTransaction()
            return item

        } catch (e: Throwable) {
            if (realm.isInTransaction) {
                realm.cancelTransaction()
            }
            e.printStackTrace()
        }

        return null
    }

    fun <T : RealmObject> findData(clazz: Class<T>): T? {

        try {
            realm.beginTransaction()
            val item = realm.where(clazz).findFirst()
            realm.commitTransaction()
            return item

        } catch (e: Throwable) {
            if (realm.isInTransaction) {
                realm.cancelTransaction()
            }
            e.printStackTrace()
        }

        return null
    }

    fun <T : RealmObject> findDataById(clazz: Class<T>, id: String): RealmResults<T> {

        return realm.where(clazz).equalTo("id", id).findAll()
    }

    fun <T : RealmObject> removeDataAll(clazz: Class<T>) {

        val result = findDataAll(clazz)
        realm.executeTransaction { result!!.deleteAllFromRealm() }
    }

    fun <T : RealmObject> removeDataById(clazz: Class<T>, id: String) {

        val result = findDataById(clazz, id)
        realm.executeTransaction { result.deleteAllFromRealm() }
    }


    //Auto Increment id
    fun <T : RealmObject> getAutoIncrementId(clazz: Class<T>): Int {

        try {
            val id = realm.where(clazz).max("id")
            return if (id != null) {
                id.toInt() + 1
            } else {
                0
            }
        } catch (e: ArrayIndexOutOfBoundsException) {
            return 0
        }

    }


}
